File: IPWorks OpenPGP Samples/OpenPGP/php/openpgp.php
<?php
/*
 * IPWorks OpenPGP 2024 PHP Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks OpenPGP in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworksopenpgp
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */
require_once('../include/ipworksopenpgp_openpgp.php');
require_once('../include/ipworksopenpgp_keymgr.php');
require_once('../include/ipworksopenpgp_const.php');
?>
<?php

class MyKeyMgr extends IPWorksOpenPGP_KeyMgr
{
  function FireKeyList($param) {
    echo $param['keyid'] . "\t" . $param['userid'] . "\tPrivate Key: " . $param['hassecretkey'] . "\n";
  }
}

try{
  if ($argc < 3) {
    echo "Usage: node openpgp.js [options]\n";
    echo "Commands: \n";
    echo "  -k, --list-keys    list keys in keyring (--homedir)\n";
    echo "  -s, --sign         sign\n";
    echo "  -e, --encrypt      encrypt data\n";
    echo "  -d, --decrypt      decrypt data\n";
    echo "  -v, --verify       verify a signature\n";
    echo "Options for key selection:\n";
    echo "  --homedir          keyring directory (containing secring.gpg and pubring.gpg)\n";
    echo "  -r, --recipient    encrypt for USER-ID\n";
    echo "  -u, --local-user   private key to use for sign or decrypt\n";
    echo "  -p,                passphrase for private key\n";
    echo "  -c,                create a test key to use\n";
    echo "Options to specify input/output:\n";
    echo "  -f                 the input file to process\n";
    echo "  -o, --output       the output file (by default, output writes to console)\n";
    echo "  -w                 whether to overwrite the output file\n";
    echo "Examples:\n";
    echo "  node openpgp.js --homedir C:/temp/keys -k                       // list keys\n";
    echo "  node openpgp.js -se --homedir C:/temp/keys -r Bob -f input.txt  // sign and encrypt for user Bob\n";
    return;
  }

  $pgp = new IPWorksOpenPGP_OpenPGP();
  $keymgr = new MyKeyMgr();
  $encrypt = $sign = $verify = $decrypt = $list = $createKey = False;
  $recipientUserId = $userId = $passphrase = $keyringDir = "";

  for ($i = 1; $i < $argc; $i++) {
    if (str_starts_with($argv[$i],"-")) {
      if ($argv[$i] == "-k" || $argv[$i] == "--list-keys") { $list = True; }
      if ($argv[$i] == "-s" || $argv[$i] == "--sign") { $sign = True; }
      if ($argv[$i] == "-e" || $argv[$i] == "--encrypt") { $encrypt = True; }
      if ($argv[$i] == "-se") { $sign = True; $encrypt = True; }
      if ($argv[$i] == "-d" || $argv[$i] == "--decrypt") { $decrypt = True; }
      if ($argv[$i] == "-v" || $argv[$i] == "--verify") { $verify = True; }
      if ($argv[$i] == "-dv") { $decrypt = True; $verify = True; }
      if ($argv[$i] == "-f" || $argv[$i] == "--file") { $pgp->setInputFile($argv[$i + 1]); }
      if ($argv[$i] == "-o" || $argv[$i] == "--output") { $pgp->setOutputFile($argv[$i + 1]); }
      if ($argv[$i] == "-w") { $pgp->setOverwrite(True); }
      if ($argv[$i] == "-r" || $argv[$i] == "--recipient") { $recipientUserId = $argv[$i + 1]; }
      if ($argv[$i] == "-u" || $argv[$i] == "--local-user") { $userId = $argv[$i + 1]; }
      if ($argv[$i] == "-p") { $passphrase = $argv[$i + 1]; }
      if ($argv[$i] == "--homedir") { $keyringDir = $argv[$i + 1]; }
      if ($argv[$i] == "-c") { $createKey = True; }
    }
  }

  if ($createKey) { // do this first, in case we need to use it
    echo "Creating a test key (test@nsoftware.com). This key will be used for all operations.\n";
    if (strlen($keyringDir) == 0) { $keyringDir = ""; }
    $keymgr->doLoadKeyring($keyringDir);
    $keymgr->doCreateKey("test@nsoftware.com", "test");
    $keymgr->doSaveKeyring($keyringDir);
    $userId = "test@nsoftware.com";
    $passphrase = "test";
    $recipientUserId = "test@nsoftware.com";
  }

  $pgp->setASCIIArmor(True);
  if ($list) {
    if (strlen($keyringDir) == 0) {
      echo "No keyring specified.\n";
      return;
    }
    echo "Listing keys in keyring: " . $keyringDir . "...\n";
    $keymgr->doLoadKeyring($keyringDir);
    $keymgr->doListKeys();
  } else {
    // check input file...
    if (strlen($pgp->getInputFile()) == 0) {
      echo "No input file specified.\n";
      return;
    }

    if ($sign && $encrypt) {
      echo "Sign and Encrypt\n";

      $pgp->setKeyCount(1);
      $pgp->setKeyKeyring(0, $keyringDir);
      $pgp->setKeyUserId(0, $userId);
      $pgp->setKeyPassphrase(0, $passphrase);

      $pgp->setRecipientKeyCount(1);
      $pgp->setRecipientKeyKeyring(0, $keyringDir);
      $pgp->setRecipientKeyUserId(0, $recipientUserId);

      $pgp->doSignAndEncrypt();
    } elseif ($sign) {
      echo "Sign\n";

      $pgp->setKeyCount(1);
      $pgp->setKeyKeyring(0, $keyringDir);
      $pgp->setKeyUserId(0, $userId);
      $pgp->setKeyPassphrase(0, $passphrase);

      $pgp->doSign();
    } elseif ($encrypt) {
      echo "Encrypt\n";

      $pgp->setRecipientKeyCount(1);
      $pgp->setRecipientKeyKeyring(0, $keyringDir);
      $pgp->setRecipientKeyUserId(0, $recipientUserId);

      $pgp->doEncrypt();
    } elseif ($decrypt && $verify) {
      echo "Decrypt and Verify\n";

      $pgp->setKeyCount(1);
      $pgp->setKeyKeyring(0, $keyringDir);
      $pgp->setKeyUserId(0, $userId);
      $pgp->setKeyPassphrase(0, $passphrase);

      $pgp->setSignerKeyCount(1);
      $pgp->setSignerKeyKeyring(0, $keyringDir);
      $pgp->setSignerKeyUserId(0, $recipientUserId);

      $pgp->doDecryptAndVerifySignature();
    } elseif ($decrypt) {
      echo "Decrypt\n";

      $pgp->setKeyCount(1);
      $pgp->setKeyKeyring(0, $keyringDir);
      $pgp->setKeyUserId(0, $userId);
      $pgp->setKeyPassphrase(0, $passphrase);

      $pgp->doDecrypt();
    } elseif ($verify) {
      echo "Verify\n";

      $pgp->setSignerKeyCount(1);
      $pgp->setSignerKeyKeyring(0, $keyringDir);
      $pgp->setSignerKeyUserId(0, $recipientUserId);

      $pgp->doVerifySignature();
    } else {
      echo "No known action specified.\n";
    }

    if ($pgp->getOutputFile() == "") {
      echo "Output Message:\n" . $pgp->getOutputMessage() . "\n";
    }

  }
} catch (Exception $e) {
  echo "Error: " . $e->getMessage() . "\n";
}
?>

File: IPWorks OpenPGP Samples/OpenPGP/python/openpgp.py
# 
# IPWorks OpenPGP 2024 Python Edition - Sample Project
# 
# This sample project demonstrates the usage of IPWorks OpenPGP in a 
# simple, straightforward way. It is not intended to be a complete 
# application. Error handling and other checks are simplified for clarity.
# 
# www.nsoftware.com/ipworksopenpgp
# 
# This code is subject to the terms and conditions specified in the 
# corresponding product license agreement which outlines the authorized 
# usage and restrictions.
# 

import sys
import string
from ipworksopenpgp import *

input = sys.hexversion<0x03000000 and raw_input or input

def ensureArg(args, prompt, index):
  if len(args) <= index:
    while len(args) <= index:
      args.append(None)
    args[index] = input(prompt)
  elif args[index] is None:
    args[index] = input(prompt)


pgp = OpenPGP()
kmgr = KeyMgr()
keyringDir = ""

def createKey():
  if(keyringDir != ""):
    listKeys()
  kmgr.create_key(input("Enter a UserId: "), input("Enter a passphrase: "))
  if(input("Create another key? (y/n): ") == "y"):
    create_key()

def fireKeyList(e):
  print("%8s  %5s            %s" %(e.key_id, e.has_secret_key, e.user_id))

def listKeys():
  print("Listing keys...")
  print("KeyId     Has Secret Key?  UserId")
  print("---------------------------------")
  kmgr.load_keyring(keyringDir)
  kmgr.list_keys()

def selectKey():
  userId = input("Enter the UserId of the private key used for signing/decryption: ")
  pgp.set_key_count(1)
  pgp.set_key_keyring(0, keyringDir)
  pgp.set_key_user_id(0, userId)
  passphrase = input("Enter the passphrase for the selected key: ")
  pgp.set_key_passphrase(0, passphrase)

def selectRecipientKey():
  userId = input("Enter the UserId of the public key used for encryption: ")
  pgp.set_recipient_key_count(1)
  pgp.set_recipient_key_keyring(0, keyringDir)
  pgp.set_recipient_key_user_id(0, userId)

def selectSignerKey():
  userId = input("Enter the UserId of the public key used for verification: ")
  pgp.set_recipient_key_count(1)
  pgp.set_recipient_key_keyring(0, keyringDir)
  pgp.set_recipient_key_user_id(0, userId)

def promptForAsciiArmor():
  return True if(input("Use ASCII Armor? (y/n): ") == "y") else False

def doEncrypt():
  selectRecipientKey()
  pgp.set_ascii_armor(promptForAsciiArmor())
  pgp.encrypt()

def doDecrypt():
  selectKey()
  pgp.decrypt()

def doSign():
  selectKey()
  pgp.set_ascii_armor(promptForAsciiArmor())
  pgp.sign()

def doVerify():
  selectSignerKey()
  pgp.verify()

def doSignAndEncrypt():
  selectKey()
  selectRecipientKey()
  pgp.set_ascii_armor(promptForAsciiArmor())
  pgp.sign_and_encrypt

def doDecryptAndVerify():
  selectKey()
  selectSignerKey()
  pgp.decrypt_and_verify()

kmgr.on_key_list = fireKeyList

print("\nWelcome to the /n software OpenPGP demo.\n")
prompt = "What would you like to do?"
prompt += "\n  0: quit"
prompt += "\n  1: encrypt"
prompt += "\n  2: decrypt"
prompt += "\n  3: sign"
prompt += "\n  4: verify"
prompt += "\n  5: sign & encrypt"
prompt += "\n  6: decrypt & verify\n"
prompt += "Choice: "
op = input(prompt)

if(op == "0"):
  sys.exit(0)

inSrc = input("Read from a file or string? (f/s): ")
if(inSrc == "s"):
  pgp.set_input_message(input("Please enter the input string: "))
else:
  pgp.set_input_file(input("Please enter the input file path: "))

outSrc = input("Write to a file or string? (f/s): ")
if(outSrc == "f"):
  pgp.set_output_file(input("Please enter the output file path: "))

keyringDir = input("Provide a path to a keyring (or for one to be created at): ")
if(input("Create a new key? (y/n): ") == "y"):
  createKey()
  kmgr.save_keyring(keyringDir)
listKeys()

if(op == "1"):
  doEncrypt()
elif(op == "2"):
  doDecrypt()
elif(op == "3"):
  doSign()
elif(op == "4"):
  doVerify()
elif(op == "5"):
  doSignAndEncrypt
elif(op == "6"):
  doDecryptAndVerify

if(outSrc == "s"):
  print("OutputMessage: %s\n" %(pgp.get_output_message()))
print("\nThe demo has finished running.")

